### ISSUE TITLE
No Attribution to the original GM_config and a change to an incompatible lincense

### ISSUE THREAD

**Issue by sizzlemctwizzle** [eight04/GM_config#1]
@eight04 
It is clear by your initial commit that you took most of the ideas behind the original [GM_config](https://github.com/sizzlemctwizzle/GM_config). Your interface is identical, but no where do you give attribution to the LGPL licensed project. You've looked at our source and then built a clone which you released under the MIT license, which is in clear violation of the LGPL. You even went further and stole the name, which belongs to me.

Please change the name of your project and comply with the terms of the LGPL.

@Martii

**Comment by Martii** [comment]
@sizzlemctwizzle 
Noted and confirmed... contact GH to have the project removed  if non-compliance including licensing within 24 hours. 

@eight04 
You should have [forked](https://github.com/sizzlemctwizzle/GM_config#fork-destination-box) the [original project](https://github.com/sizzlemctwizzle/GM_config) and kept the licensing intact.

**Comment by eight04** [comment]
Sorry for the stealing. I am not that familier with licenses. 
- This project was inspired from a script called [GM_config (JoeSimmons)](https://greasyfork.org/en/scripts/1884-gm-config/code).
- I use same api interface with it. (`GM_config.init/GM_config.open`)
- Should I keep the original license even all codes are written from scratch?
- If so, should I use the license belongs to original project or that script?

**Comment by Martii** [comment]
@sizzlemctwizzle is the establishing author. You are welcome to fork the project appropriately here on GitHub for attribution and contribute to your own fork of it with whatever changes. As for Joe's version, which is also a fork, one of his homes is [here](https://openuserjs.org/scripts/JoeSimmons/GM_config).

License has to stay LGPL and the header has to stay in the files and should be propagated to each derivation.

**Comment by sizzlemctwizzle** [comment]
Yes JoeSimmons has a fork which is [properly licensed](https://greasyfork.org/en/scripts/1884-gm-config/code#n10).

It doesn't matter if you wrote it from scratch. You clearly looked at the source (some evidence is that you also fallback to localStorage when GM_set/getValue are unavailable, and use an internal element creation function) and used that code for inspiration, which makes it a derivative.

> If so, should I use the license belongs to original project or that script?

Both are licensed under the LGPL, and your derivative must be as well. I don't care if you don't directly fork the original project on GitHub. Just that you copy [the header](https://github.com/sizzlemctwizzle/GM_config/blob/master/gm_config.js#L1).

**Comment by eight04** [comment]
Got it. I will fix it later. Thanks for the reply.

**Comment by sizzlemctwizzle** [comment]
I would rather that you chose a different name since you essentially rewrote the library (mostly to prevent confusion). But I'll settle for you pointing to https://github.com/sizzlemctwizzle/GM_config and stating that your project is a complete rewrite.

**Comment by Martii** [comment]
Still has to be LGPL though with headers.

**Comment by eight04** [comment]
I have changed lheader and readme description, also pulled license files from original repo.
The [script on Greasyfork](https://greasyfork.org/zh-TW/scripts/7212-gm-config-eight-s-version) should be automatically synced in a minute.

**Comment by sizzlemctwizzle** [comment]
> This project is a rewrite of sizzlemctwizzle/GM_config and use the interface similar to GM_config (JoeSimmons).

The interface JoeSimmons' fork uses is the legacy interface of the official GM_config (it is still supported but not documented).

**Comment by Martii** [comment]
@eight04 

btw the LGPL header is currently this:

``` javascript
/*
Copyright 2009+, GM_config Contributors (https://github.com/sizzlemctwizzle/GM_config)

GM_config Contributors:
    Mike Medley <medleymind@gmail.com>
    Joe Simmons
    Izzy Soft
    Marti Martz

GM_config is distributed under the terms of the GNU Lesser General Public License.

    GM_config is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
```

The UserScript metadata block key you put in of `@license` is a start on the userscript side but you need to keep the license header in place before the metadata block. You can add your own [`@copyright`](http://sourceforge.net/p/greasemonkey/wiki/Metadata_Block/#copyright) for your changes to the UserScript metadata block but the project needs that unmodified attribution header.

**Comment by eight04** [comment]
@Martii 

I have changed the folder structure, hope I am not missing something.
- The js source file located at [src/GM_config.js](https://github.com/eight04/GM_config/blob/master/src/GM_config.js), which includes `@license`, `@attribution` meta keys.
- The license file located in [license](https://github.com/eight04/GM_config/tree/master/license) folder, including LGPL license and the header as _copyright.txt_.
- Build the source to [dist/GM_config.js](https://github.com/eight04/GM_config/blob/master/dist/GM_config.js).

> you need to keep the license header in place before the metadata block.

Does it mean that I have to concat license header before metadata block? What about to put the header between the source and metadata block? The script parser might not be able to parse it if `// ==UserScript==` does not present at the first line.

P.S. I never know that there are so many meta keys. Thanks Martii!

**Comment by Martii** [comment]
> The script parser might not be able to parse it if `// ==UserScript==` does not present at the first line.

The UserScript metadata block specification is a serialized stream anywhere in a user.js file and can be considered independent of Code contained therein... if you get bored you can put the UserScript metadata block at the end of the file or even split it up _(some at the beginning, some in the middle, and some at the end)_... so all user.js engines should always be able to read it... including all user.js hosting sites like https://openuserjs.org/ does as well.  Some examples of splits interlaced between Code are at [Knowing your own metadata](http://sourceforge.net/p/greasemonkey/wiki/Metadata_Block/#knowing-your-own-metadata).

With standardized GPL and LGPL the header should be included in all subsequent derivatives even if portions are split up code wise. GM_config _(GMC)_ is simple enough that it really only needs the license header included... if the target project is utilizing `gm_config.js`, which is a library script btw not necessarily a user.js, then the license header needs to be included. The UserScript metadata block that Aaron _(creator of Greasemonkey a.k.a. GM)_ made deals with distinct keys to achieve similar goals... however licensing terms with GPL and LGPL states that at minimum the header should be in all files derived. Some of the non-recognized and ignored keys are usually meant for user.js and **not** library .js files... there is however a little bit of cross-over.

With licensing all Copyrights should remain intact for the original authors... your Copyright as well but distinctly different. With GMC it acknowledges attribution of the creators, in which I'm a contributor primarily with some cleanup and an occasional bug fix but Sizzle, Joe, and Izzy are the main original authors... so their Copyright always has to be recognized in every derivative.

Header files are designed to keep things smaller as you will notice the last sentence says "You should have received a copy of the GNU Lesser General Public License
along with this program. If not, see http://www.gnu.org/licenses/." instead of pages and pages of legal preamble... but the Copyrights remain intact. In case upstream GMC doesn't want to support something that you do this limits everyones liability and allows it continue as free and open source through all propagated derivatives.

Most of your tree now looks quite good however https://github.com/eight04/GM_config/blob/master/src/GM_config.js still has to have the header in it in case it is directly `@require`d by a top level user.js . Once that is done you can probably close this issue. @sizzlemctwizzle is a little less of a stickler but the community around GPL and LGPL _(L/GPL)_ are quite picky in order to keep free software free. :) 

Hope some of this banter helps.

**Comment by eight04** [comment]
So if I am going to split _GM_config.js_ to different part (e.g. _dialogManager.js_, _interface.js_, ...), I have to include the header to each part right?

**Comment by Martii** [comment]
> So if I am going to split _GM_config.js_ to different part (e.g. dialogManager.js, interface.js, ...), I have to include the header to each part right?

Correct according to L/GPL legalese.
